python/lib/scheduler/dmod/scheduler/resources/resource_manager.py

```diff
@@ -170,126 +184,177 @@ def allocate_single_node(self, cpus: int, memory: int) -> List[Optional[Resource
-            with up to num_node-1 remainders to fill in.  This is not the most generic "round robin" in which we allocate
+            Generate even, balanced allocations of the requested assets across all nodes that are active and ready.
+            Importantly, this is slightly different not just those within
+
+            This is a balanced round robin algorithm, assuming an even distribution is possible across all resources,
+            with up to num_node-1 remainders to fill in.
+            Note that this is not the most generic "round robin" in which we allocate
             cpus one after the other across all available resources and don't try to balance.
             i.e. a request for 10 cpus with an available resource view of [4, 2, 4] would fail to allocate with this
-            algorithm, because it assumes an availablity of [4, 3, 3]
+            algorithm, because it assumes an availability of [4, 3, 3]
+
+            For a ``BUNDLED`` :class:`AllocationAssetGrouping`, a single allocation is returned per :class:`Resource`
+            node containing all the assets allocated from that node. For ``SILO``, several per-node allocations are
+            returned, each with a single CPU and a roughly even share of the total requested memory.
 
             Parameters
             ----------
                 cpus: Total number of CPUs requested
                 memory: Amount of memory required in bytes
+                asset_grouping: The way compute assets from a single node are grouped into allocations
 
             Returns
             -------
             [ResourceAlloction]
                 List of one or more ResourceAllocation if allocation successful, otherwise, [None]
         """
         #TODO consider scaling memory per cpu
-        #Find the number of cpus to allocate to each node
         self.validate_allocation_parameters(cpus, memory)
-        resources = list(self.get_useable_resources())
-        per_alloc_mem = int(memory / cpus)
-
-        num_node = len(resources)
-        if num_node == 0:
+        # Get all active and ready nodes, regardless of whether they are "allocateable" (i.e., not full)
+        resource_nodes = {r.resource_id: r for r in self.get_resources() if r.is_active() and r.is_ready()}
+
+        # Calculate in advance the exact amounts of CPUs and memory per node for the necessary balance
+        # This is slightly different from simply an even share due to discrete amounts and remainders
+        cpus_per_node, memory_per_node = dict(), dict()
+        num_nodes = len(resource_nodes)
+        cpu_share, cpu_remainder = divmod(cpus, num_nodes)
+        mem_share, mem_remainder = divmod(memory, num_nodes)
+
+        for node_id, node in resource_nodes.items():
+            # We must plan to request at least the per-node amounts on each node
+            if node.cpu_count < cpu_share or node.memory < memory:
```

Comment:
This should be `mem_share` right?
```suggestion
            if node.cpu_count < cpu_share or node.memory < mem_share:
```

Reply:
I think we should pull this `for` loop out into its own function similar to `validate_allocation_parameters` to verify if _we_ can schedule something or not. The function should not depend on IO, but just be passed what is needed and ideally we could limit the scope of the input parameters using a `typing.Protocol` to make it really easy to test this.

Reply:
Moving the `for` loop seems excessive to me.  It's a relatively small loop with logic only applicable to the function it's in.  `validate_allocation_parameters` is used in a few places.   Maybe I'm not following exactly what you're trying to say here.